## Cho tập tin "shark.tsv" chứa dữ liệu số vụ cá mập tấn công các tháng trong năm của 3 quốc gia USA, Australia, South Africa)

In [63]:
import numpy as np

In [64]:
#1. Đọc dữ liệu từ tập tin shark.tsv và đưa vào mảng shark(Gợi ý: sử dụng hàm np.genfromtxt) 
shark = np.genfromtxt('shark.tsv', delimiter = '\t', skip_header= 1, filling_values= -1, dtype=int)
shark

array([[ -1,  28,  94,  68],
       [ -1,  27,  78,  32],
       [ -1,  66,  63,  34],
       [ -1, 103,  54,  25],
       [ -1, 106,  21,  25],
       [ -1, 173,  25,  22],
       [ -1, 232,  25,  28],
       [ -1, 208,  28,  15],
       [ -1, 182,  35,  17],
       [ -1, 141,  56,  16],
       [ -1,  79,  65,  20],
       [ -1,  37,  94,  46]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[ -1,  28,  94,  68],
       [ -1,  27,  78,  32],
       [ -1,  66,  63,  34],
       [ -1, 103,  54,  25],
       [ -1, 106,  21,  25],
       [ -1, 173,  25,  22],
       [ -1, 232,  25,  28],
       [ -1, 208,  28,  15],
       [ -1, 182,  35,  17],
       [ -1, 141,  56,  16],
       [ -1,  79,  65,  20],
       [ -1,  37,  94,  46]])
</pre>

</details>

In [65]:
#2. Cho biết số phần tử mỗi chiều của mảng
shark.shape

(12, 4)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>(12, 4)
</pre>

</details>

In [66]:
#3. Thay cột 1 (giá trị -1) bằng giá trị tương ứng các tháng trong năm 1-12
shark[:,0] = np.arange(1,13)
shark

array([[  1,  28,  94,  68],
       [  2,  27,  78,  32],
       [  3,  66,  63,  34],
       [  4, 103,  54,  25],
       [  5, 106,  21,  25],
       [  6, 173,  25,  22],
       [  7, 232,  25,  28],
       [  8, 208,  28,  15],
       [  9, 182,  35,  17],
       [ 10, 141,  56,  16],
       [ 11,  79,  65,  20],
       [ 12,  37,  94,  46]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[  1,  28,  94,  68],
       [  2,  27,  78,  32],
       [  3,  66,  63,  34],
       [  4, 103,  54,  25],
       [  5, 106,  21,  25],
       [  6, 173,  25,  22],
       [  7, 232,  25,  28],
       [  8, 208,  28,  15],
       [  9, 182,  35,  17],
       [ 10, 141,  56,  16],
       [ 11,  79,  65,  20],
       [ 12,  37,  94,  46]])
</pre>

</details>

In [67]:
#4. Hiển thị số vụ cá mập tấn công trung bình của 3 quốc gia
np.mean(shark[:,1:],axis = 0)

array([115.16666667,  53.16666667,  29.        ])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([115.16666667,  53.16666667,  29.        ])
</pre>

</details>

In [68]:
#5. Hiển thị số vụ cá mập tấn công trung bình của mỗi tháng  
np.mean(shark[:,1:], axis = 1)

array([63.33333333, 45.66666667, 54.33333333, 60.66666667, 50.66666667,
       73.33333333, 95.        , 83.66666667, 78.        , 71.        ,
       54.66666667, 59.        ])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([63.33333333, 45.66666667, 54.33333333, 60.66666667, 50.66666667,
       73.33333333, 95.        , 83.66666667, 78.        , 71.        ,
       54.66666667, 59.        ])
</pre>

</details>

In [69]:
#6. Tạo thêm cột tổng số vụ cá mập tấn công theo mỗi tháng 
col_sum=np.sum(shark[:,1:], axis = 1)
col_sum

array([190, 137, 163, 182, 152, 220, 285, 251, 234, 213, 164, 177])

In [70]:
shark = np.concatenate([shark, col_sum.reshape(12,1)], axis=1)
shark

array([[  1,  28,  94,  68, 190],
       [  2,  27,  78,  32, 137],
       [  3,  66,  63,  34, 163],
       [  4, 103,  54,  25, 182],
       [  5, 106,  21,  25, 152],
       [  6, 173,  25,  22, 220],
       [  7, 232,  25,  28, 285],
       [  8, 208,  28,  15, 251],
       [  9, 182,  35,  17, 234],
       [ 10, 141,  56,  16, 213],
       [ 11,  79,  65,  20, 164],
       [ 12,  37,  94,  46, 177]])

In [71]:
np.insert(shark, 4, col_sum.T, axis = 1)

array([[  1,  28,  94,  68, 190, 190],
       [  2,  27,  78,  32, 137, 137],
       [  3,  66,  63,  34, 163, 163],
       [  4, 103,  54,  25, 182, 182],
       [  5, 106,  21,  25, 152, 152],
       [  6, 173,  25,  22, 220, 220],
       [  7, 232,  25,  28, 285, 285],
       [  8, 208,  28,  15, 251, 251],
       [  9, 182,  35,  17, 234, 234],
       [ 10, 141,  56,  16, 213, 213],
       [ 11,  79,  65,  20, 164, 164],
       [ 12,  37,  94,  46, 177, 177]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[  1,  28,  94,  68, 190],
       [  2,  27,  78,  32, 137],
       [  3,  66,  63,  34, 163],
       [  4, 103,  54,  25, 182],
       [  5, 106,  21,  25, 152],
       [  6, 173,  25,  22, 220],
       [  7, 232,  25,  28, 285],
       [  8, 208,  28,  15, 251],
       [  9, 182,  35,  17, 234],
       [ 10, 141,  56,  16, 213],
       [ 11,  79,  65,  20, 164],
       [ 12,  37,  94,  46, 177]])
</pre>

</details>

In [72]:
#7. Cho biết những tháng có tổng số vụ cá mập tấn công >200
shark[shark[:,4] > 200]


array([[  6, 173,  25,  22, 220],
       [  7, 232,  25,  28, 285],
       [  8, 208,  28,  15, 251],
       [  9, 182,  35,  17, 234],
       [ 10, 141,  56,  16, 213]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[  6, 173,  25,  22, 220],
       [  7, 232,  25,  28, 285],
       [  8, 208,  28,  15, 251],
       [  9, 182,  35,  17, 234],
       [ 10, 141,  56,  16, 213]])
</pre>

</details>

In [73]:
#8. Cho biết tổng số vụ cá mập tấn công nhiều nhất trong tháng là bao nhiêu
max = shark[:,4].max()
max

np.int64(285)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>285
</pre>

</details>

In [74]:
#9. Cho biết tháng nào có tổng số vụ cá mập tấn công nhiều nhất
month = shark[shark[:,-1] == max][0,0]
print(f'Tháng có tổng số vụ cá mập tấn công nhiều nhất là tháng : {month}')


Tháng có tổng số vụ cá mập tấn công nhiều nhất là tháng : 7


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>Tháng có tổng số vụ cá mập tấn công nhiều nhất là tháng : 7
</pre>

</details>